src/coreclr/tools/superpmi/superpmi-shared/compileresult.cpp

```diff
@@ -891,12 +892,143 @@ void CompileResult::applyRelocs(unsigned char* block1, ULONG blocksize1, void* o
+                            if (index == -1)
+                            {
+                                // See if the original address is in the replay address map. This happens for
+                                // relocations on static field addresses found via getFieldAddress().
+                                void* origAddr = repAddressMap((void*)tmp.target);
+                                if ((origAddr != (void*)-1) && (origAddr != nullptr))
+                                {
+                                    key   = CastPointer(origAddr);
+                                    index = rc->mc->GetRelocTypeHint->GetIndex(key);
+                                    if (index != -1)
+                                    {
+                                        LogDebug("    Using address map: target %016llX, original target %016llX",
+                                            tmp.target, key);
+                                    }
+                                }
+                            }
+
+                            if (index != -1)
+                            {
+                                WORD retVal = (WORD)rc->mc->GetRelocTypeHint->Get(key);
+                                if (retVal == IMAGE_REL_BASED_REL32)
+                                {
+                                    LogDebug("    REL32 target used as argument to getRelocTypeHint: setting delta=%d (0x%X)",
+                                             (int)key, (int)key);
+                                    delta        = (INT64)(int)key;
+                                    deltaIsFinal = true;
+                                }
+                            }
+                        }
+                    }
+
+                    if (!deltaIsFinal)
+                    {
+                        // Check if tmp.target is the result of a call to getHelperFtn(). If so, the VM would create a
+                        // jump stub if the REL32 address doesn't fit. We don't want to fail with a REL32 overflow if
+                        // the actual target address doesn't fit, so use the low-order 32 bits of the address.
+                        // We need to iterate the entire table since we don't know the helper function id.
+
+                        if (rc->mc->GetHelperFtn != nullptr)
+                        {
+                            for (unsigned int idx = 0; idx < rc->mc->GetHelperFtn->GetCount(); idx++)
+                            {
+                                DLDL value = rc->mc->GetHelperFtn->GetItem(idx);
+                                if (value.B == tmp.target)
+                                {
+                                    LogDebug("    REL32 target is result of getHelperFtn(): setting delta=%d (0x%X)",
+                                             (int)tmp.target, (int)tmp.target);
+                                    delta        = (INT64)(int)tmp.target;
+                                    deltaIsFinal = true;
```

Comment:
shouldn't we `break` from the for loops (here and below) once we set `deltaIsFinal = true`?

Reply:
No. The `for` loop is over all the relocations to be applied; `deltaIsFinal` is within a single relocation handling